dGame/Entity.cpp

```diff
@@ -1225,18 +1225,28 @@ void Entity::UpdateXMLDoc(tinyxml2::XMLDocument* doc) {
 
 void Entity::Update(const float deltaTime) {
 	uint32_t timerPosition;
-	for (timerPosition = 0; timerPosition < m_Timers.size();) {
+	size_t currentSize = m_Timers.size();
+	for (timerPosition = 0; timerPosition < currentSize;) {
 		auto& timer = m_Timers[timerPosition];
 		timer.Update(deltaTime);
 		// If the timer is expired, erase it and dont increment the position because the next timer will be at the same position.
-		// Before: [0, 1, 2, 3, ..., n]
+		// Before: [0, 1, 2, 3, ..., sizeBeforeLoop, newTimers]
 		// timerPosition  ^
-		// After:  [0, 1, 3, ..., n]
+		// currentSize               ^^^^^^^^^^^^^^
+		// After:  [0, 1, n, ..., sizeBeforeLoop - 1, newTimers] 2 is expired and removed now and currentSize still points at the end of the vector before the loop started
 		// timerPosition  ^
+		// currentSize            ^^^^^^^^^^^^^^^^^^
 		if (timer.GetTime() <= 0) {
 			// Remove the timer from the list of timers first so that scripts and events can remove timers without causing iterator invalidation
 			auto timerName = timer.GetName();
-			m_Timers.erase(m_Timers.begin() + timerPosition);
+			// We don't need to copy the element if there is only 1 element nor do we need to copy if we are on the last element.
+			// This is a clever removal trick that avoids having to copy the entire vector and instead replaces this now expired
+			// element with the last element in the vector and then removes the last element.
+			currentSize--;
+			if (currentSize > 1 && timerPosition < currentSize) {
+				m_Timers.at(timerPosition) = m_Timers.at(currentSize);
+			}
+			m_Timers.erase(m_Timers.begin() + currentSize);
```

Comment:
Since that is no longer real size - 1, it'll move all the new timers unless you do a second swap, right?

Reply:
> Since that is no longer real size - 1, it'll move all the new timers unless you do a second swap, right?

no?  on erase, the only thing that happens is 1 copy and 1 shrink before any additions take place.

Reply:
Can you elaborate on that?

Reply:
im not sure what a second swap would even be for.  the only swapping and moving is anything that happens to the left of the added timers.  Added timers are not touched this frame.  Any moving/copying/swapping is only happening if the container grows in size and re-allocates.